src/main/kotlin/persistence/repo/UserRepo.kt

```diff
@@ -0,0 +1,132 @@
+     */
+    override fun insert(user: User): Observable<Int> {
+        // creates observable to return generated int
+        return Observable.create<Int> {
+            it.onNext(dataStore.insert(userMapper.mapToEntity(user)).id)
+        }.doOnNext {
+            updateUserLanguageReferences(user, it)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets user by Id
+     */
+    override fun getById(id:Int): Observable<User>{
+        return Observable.create<IUserEntity> {
+            it.onNext(dataStore.select(IUserEntity::class)
+                    .where(IUserEntity::id eq id)
+                    .get().first())
+        }.map {
+            userMapper.mapFromEntity(it)
+        }.subscribeOn(Schedulers.io())
+
+    }
+
+    /**
+     * given a audioHash gets the user
+     */
+    fun getByHash(hash: String): Observable<User> {
+        return Observable.create<IUserEntity> {
+            dataStore.select(IUserEntity::class)
+                    .where(IUserEntity::audioHash eq hash)
+                    .get().first()
+        }.map {
+            userMapper.mapFromEntity(it)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets all the users currently stored in db
+     */
+    override fun getAll(): Observable<List<User>>{
+        return Observable.create<List<IUserEntity>> {
+            it.onNext(dataStore.select(IUserEntity::class)
+                    .get().toList())
+        }.map {
+            it.map { userMapper.mapFromEntity(it) }
+        }.subscribeOn(Schedulers.io())
+
+    }
+
+    override fun update(user: User): Completable {
+        return Completable.fromAction {
+                val userEntity = userMapper.mapToEntity(user)
+                dataStore.update(userEntity)
+                dataStore.update(userEntity.userPreferencesEntity)
+                updateUserLanguageReferences(user, user.id)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * deletes user by id
+     */
+    override fun delete(user: User): Completable{
+        return Completable.fromAction {
+            val userEntity = userMapper.mapToEntity(user)
+            dataStore.delete(userEntity)
+            dataStore.delete(IUserLanguage::class).where(IUserLanguage::userEntityid eq user.id).get().value()
+        }.subscribeOn(Schedulers.io())
+    }
+
+    private fun updateUserLanguageReferences(user: User, userId: Int) {
+        // inserts source and target languages into user language relationship table
+        val newSourceUserLanguages = user.sourceLanguages.map {
+            val tmp = UserLanguage()
+            tmp.setLanguageEntityid(it.id)
+            tmp.setUserEntityid(userId)
+            tmp.setSource(true)
+            tmp
+        }
+        val newTargetUserLanguages = user.targetLanguages.map {
+            val tmp = UserLanguage()
+            tmp.setLanguageEntityid(it.id)
+            tmp.setUserEntityid(userId)
+            tmp.setSource(false)
+            tmp
```

Comment:
Can we use something other than `tmp` here?

Reply:
yes